### PR TITLE
feat: Document Service 

### PR DESCRIPTION
**Pull request by Marc-Roig**
NOTE from @innerdvations : This is not fully implemented and contains many known bugs, but it's now in a state where the core functionality is present and we plan to merge it in to `v5/main` Monday Nov 27 so that it can start being used and tested. Please add review comments but be aware they will be added to the backlog and addressed in later PRs.

### What does it do?

Implementation of the document service.

You can test the document service like:

```ts
	// FIND MANY
    const findManyResult = await strapi.documentService.findMany('api::restaurant.restaurant');
    // FIND ONE
    const findOneResult = await strapi.documentService.findOne('api::restaurant.restaurant', DOC_ID);
    // FIND FIRST
    const findFirstResult = await strapi.documentService.findFirst('api::restaurant.restaurant', {
      filters: { name: { $startsWith: 'A' } },
    });
    // CREATE
    const createResult = await strapi.documentService.create('api::restaurant.restaurant', {
      data: { name: 'A new restaurant' },
    });
    // UPDATE
    const updateResult = await strapi.documentService.update(
      'api::restaurant.restaurant',
      DOC_ID,
      {
        data: { name: 'Other Restaurant' },
        filters: { name: { $startsWith: 'M' } },
      }
    );
	// DELETE
    const deletedRestaurant = await strapi.documentService.delete('api::restaurant.restaurant', DOC_ID);
```

```
- Tests
- Components
- Other methods
	-  Count
	-  FindPage
	-  Clone
	-  Load
	-  LoadPages
	-  DeleteMany
- Entity Validation
- Sanitization
- Lifecycles
-  Plugin extensions
- Apply default parameters (status & locale)
- D&P
- i18n
- Webhooks
- Audit logs
- File upload
- Transactions. Let's do proper testing.
- Documentation
```

### CODE REVIEW

File: packages/core/strapi/src/services/document-service/document-repository.ts [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+        return middlewareManager.run(
+          {
+            action: 'findOne',
+            uid,
+            params,
+            options: { id },
+          },
+          ({ params }) => documents.findOne(uid, id, params)
+        );
+      },
```
Comment (Marc-Roig) [inline packages/core/strapi/src/services/document-service/document-repository.ts:46]: @alexandrebodin we talked about using a "compose" pattern to run middlewares.

I am not able to find something that makes it clear 🤔 
```
      findOne: compose(
        middlewareManager.getFor(uid, 'findOne'), 
        () =>documents.findOne(uid, id, params)
      )({
        ctx: {
          action: 'findOne',
          uid,
          params,
          options: { id },
        },
      }),
```

I abstracted away the middleware registration into `middlewareManager`, I don't know if there are other alternatives that could help here.

Do you have some examples of how we use the compose pattern in our repo? I only found one simple example with a hardcoded middleware.

File: packages/core/strapi/src/services/document-service/index.ts [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+      .query(uid)
+      .findMany({ ...query, where: { ...params?.lookup, ...query?.where, documentId } });
+
+    // Document does not exist
+    if (!entries.length) {
+      return null;
+    }
+
+    // TODO: How do we do this from i18n package?
+    const entryToUpdate = entries.find((entry) => entry.locale === params?.locale);
```
Comment (Marc-Roig) [inline packages/core/strapi/src/services/document-service/index.ts:176]: I wonder how we could handle this in a clean way. I don't believe the logic to know if an entry has to be updated or created in a document should be in i18n.

**Review by innerdvations — APPROVED** [review]
Forgot to post this review yesterday, but doing it now so that we have a record for additional PRs in the next few days.

File: packages/core/strapi/src/services/document-service/document-repository.ts [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+ *    return next(ctx)
+ *  }
+ *
+ *  // Middleware for specific uid
+ *  strapi.documents.use('api::article.article', async (ctx, next) => {
+ *    // Filter private fields
+ *    const { privateField, ...result } = await next(ctx)
+ *    return result;
+ *
+ *  })
```
Comment (remidej) [inline packages/core/strapi/src/services/document-service/document-repository.ts]: @Marc-Roig don't you need to specify the action here? e.g. `findMany`

**Pull request merged**